src/api/c/pinverse.cpp

```diff
@@ -0,0 +1,199 @@
+/*******************************************************
+ * Copyright (c) 2018, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <backend.hpp>
+#include <Array.hpp>
+
+#include <af/array.h>
+#include <af/complex.h>
+#include <af/defines.h>
+#include <af/lapack.h>
+#include <arith.hpp>
+#include <blas.hpp>
+#include <cast.hpp>
+#include <common/ArrayInfo.hpp>
+#include <common/err_common.hpp>
+#include <diagonal.hpp>
+#include <handle.hpp>
+#include <logic.hpp>
+#include <math.hpp>
+#include <reduce.hpp>
+#include <select.hpp>
+#include <svd.hpp>
+#include <tile.hpp>
+#include <transpose.hpp>
+
+using af::dim4;
+using af::dtype_traits;
+using std::vector;
+using std::swap;
+
+using namespace detail;
+
+const double dfltTol = 1e-6;
+
+template<typename T>
+Array<T> getSubArray(const Array<T> &in, const bool copy,
+                       uint dim0begin = 0, uint dim0end = 0,
+                       uint dim1begin = 0, uint dim1end = 0,
+                       uint dim2begin = 0, uint dim2end = 0,
+                       uint dim3begin = 0, uint dim3end = 0) {
+    vector<af_seq> seqs = {
+        {static_cast<double>(dim0begin), static_cast<double>(dim0end), 1.},
+        {static_cast<double>(dim1begin), static_cast<double>(dim1end), 1.},
+        {static_cast<double>(dim2begin), static_cast<double>(dim2end), 1.},
+        {static_cast<double>(dim3begin), static_cast<double>(dim3end), 1.}
+    };
+    return createSubArray<T>(in, seqs, copy);
+}
+
+// Moore-Penrose Pseudoinverse
+template<typename T>
+Array<T> pinverseSvd(const Array<T> &in, const double tol)
+{
+    in.eval();
+    int M = in.dims()[0];
+    int N = in.dims()[1];
+    int P = in.dims()[2];
+    int Q = in.dims()[3];
+
+    // Compute SVD
+    typedef typename dtype_traits<T>::base_type Tr;
+    // Ideally, these initializations should use createEmptyArray(), but for some
+    // reason, linux-opencl-k80 will produce wrong results for large arrays
+    Array<T> u = createValueArray<T>(dim4(M, M, P, Q), scalar<T>(0));
+    Array<T> vT = createValueArray<T>(dim4(N, N, P, Q), scalar<T>(0));
+    Array<Tr> sVec = createValueArray<Tr>(dim4(min(M, N), 1, P, Q), scalar<Tr>(0));
```

Comment:
I understand the reason behind using createValueArray, but wondering if it justifies having three memsets because of that.

Does this happen only on k80 with OpenCL backend ?

Reply:
Yes it only happens on the K80 with the OpenCL backend. I wish I knew how to solve it another way.

Reply:
I get the situation, lets deal with a work-around if at all possible later.